File: 2015/day8/day8.py
def calculate_difference(input_file):
    total_code_count = 0
    total_memory_count = 0

    with open(input_file, 'r') as file:
        for line in file:
            code_count = len(line.strip())
            line = line[1:-2]  # Remove surrounding double quotes
            memory_count = 0

            i = 0
            while i < len(line):
                if line[i] == '\\':
                    i += 1
                    if line[i] in ['\\', '\"']:
                        memory_count += 1
                    elif line[i] == 'x':
                        memory_count += 1
                        i += 2
                else:
                    memory_count += 1
                i += 1

            total_code_count += code_count
            total_memory_count += memory_count

    difference = total_code_count - total_memory_count

    print("Total Code Count:", total_code_count)
    print("Total Memory Count:", total_memory_count)
    print("Difference:", difference)


def encode_strings(input_file):
    total_code_count = 0
    total_encoded_count = 0

    with open(input_file, 'r') as file:
        for line in file:
            code_count = len(line.strip())
            encoded_string = "\"" + "".join(["\\" + c if c in ['\\', '\"'] else c for c in line.strip()]) + "\""
            encoded_count = len(encoded_string)

            total_code_count += code_count
            total_encoded_count += encoded_count

    difference = total_encoded_count - total_code_count

    print("Total Code Count:", total_code_count)
    print("Total Encoded Count:", total_encoded_count)
    print("Difference:", difference)


# Example usage:
input_file = "instructions.txt"
print("Part One:")
calculate_difference(input_file)

print("\nPart Two:")
encode_strings(input_file)


    

File: 2015/day8/day8.cpp
#include <iostream>
#include <fstream>
#include <string>

int main() {
    std::ifstream inputFile("instructions.txt");
    if (!inputFile.is_open()) {
        std::cerr << "Error opening file." << std::endl;
        return 1;
    }

    std::string line;
    int totalCodeCount = 0;
    int totalMemoryCount = 0;

    while (std::getline(inputFile, line)) {
        int codeCount = line.size();

        line = line.substr(1, line.size() - 2);

        int memoryCount = 0;
        for (size_t i = 0; i < line.size(); ++i) {
            if (line[i] == '\\') {
                ++i; 
                if (line[i] == '\\' || line[i] == '\"') {
                    memoryCount += 1;
                } else if (line[i] == 'x') {
                    memoryCount += 1;
                    i += 2; 
                }
            } else {
                memoryCount += 1;
            }
        }

        totalCodeCount += codeCount;
        totalMemoryCount += memoryCount;
    }

    inputFile.close();

    int difference = totalCodeCount - totalMemoryCount;

    std::cout << "Total Code Count: " << totalCodeCount << std::endl;
    std::cout << "Total Memory Count: " << totalMemoryCount << std::endl;
    std::cout << "Difference: " << difference << std::endl;

    return 0;
}
